src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs

```diff
@@ -406,147 +391,158 @@ private static int BinarySearch(T[] array, int index, int length, T value)
             return ~lo;
         }
 
-        private static void SwapIfGreater(Span<T> keys, int i, int j)
+        /// <summary>Swaps the values in the two references if the references are not identical and the first is greater than the second.</summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void SwapIfGreater(ref T i, ref T j)
         {
-            Debug.Assert(0 <= i && i < keys.Length);
-            Debug.Assert(0 <= j && j < keys.Length);
-
-            if (i != j)
+            if (!Unsafe.AreSame(ref i, ref j))
             {
-                if (keys[i] != null && keys[i].CompareTo(keys[j]) > 0)
-                {
-                    T key = keys[i];
-                    keys[i] = keys[j];
-                    keys[j] = key;
-                }
+                SwapIfGreaterKnownDistinct(ref i, ref j);
             }
         }
 
+        /// <summary>Swaps the values in the two references if the first is greater than the second.</summary>
+        /// <remarks>No check is performed for whether the references are the same, so additional overhead is incurred swapping unnecessarily if they are.</remarks>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static void Swap(Span<T> a, int i, int j)
+        private static void SwapIfGreaterKnownDistinct(ref T i, ref T j)
         {
-            if (i != j)
+            if (i != null && i.CompareTo(j) > 0)
             {
-                T t = a[i];
-                a[i] = a[j];
-                a[j] = t;
+                SwapKnownDistinct(ref i, ref j);
             }
         }
 
-        internal static void IntrospectiveSort(Span<T> keys)
+        /// <summary>Swaps the values in the two references if the references are not identical.</summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Swap(ref T i, ref T j)
         {
-            if (keys.Length > 1)
+            if (!Unsafe.AreSame(ref i, ref j))
             {
-                IntroSort(keys, 2 * IntrospectiveSortUtilities.FloorLog2PlusOne(keys.Length));
+                SwapKnownDistinct(ref i, ref j);
             }
         }
 
+        /// <summary>Swaps the values in the two references, regardless of whether the two references are the same.</summary>
+        /// <remarks>No check is performed for whether the references are the same, so additional overhead is incurred swapping unnecessarily if they are.</remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void SwapKnownDistinct(ref T i, ref T j)
+        {
+            Debug.Assert(!Unsafe.AreSame(ref i, ref j));
+            T t = i;
```

Comment:
Does/can the JIT optimize this to just an `xchg` instruction?

Reply:
> Does/can the JIT optimize this to just an xchg instruction?

It doesn't, at least not from what I saw.

Reply:
Maybe another potential optimization here, might investigate a bit more and log a bug.

Reply:
> might investigate a bit more and log a bug

To clarify, you might, or you're suggesting I do?  If you have the time, that'd be great :smile:

Reply:
I was saying I might. Its definitely not a blocker on this or anything.
Just saw the code and realized I've seen the same pattern elsewhere and was thinking it was a potential optimization we might be missing.

Reply:
Looks like its not actually recommended due to a huge latency for `xchg reg, [mem]`, so none of the native compilers use it.